proposals/0000-add-support-for-css-color-function.md

```diff
@@ -0,0 +1,109 @@
+---
+title: Add support for CSS color() function
+author:
+  - Ryan Linton
+  - Yulian Glukhenko
+date: 2023-11-20
+---
+
+# RFC0000: Add support for CSS color() function
+
+## Summary
+
+React Native does [not currently support](https://github.com/facebook/react-native/issues/41517) wide color gamut color spaces (i.e. display-p3). This proposal discusses adding support for the CSS color() function enabling support for additional wide-gamut color spaces.
+
+## Basic example
+
+```js
+StyleSheet.create({
+  backgroundColor: "color(display-p3 1 0.5 0)",
+  color: "color(display-p3 1 0.5 0 / .5)",
+});
+```
+
+## Motivation
+
+Since most new devices support wider gamut color spaces, React Native should support them as well. The Display P3 color space has had native support since Android 8.0 and iOS 9.3. The color() function was introduced with [CSS Color Module Level 4](https://drafts.csswg.org/css-color/#color-function), much of which is already implemented in React Native. Also Flutter [recently added support](https://github.com/flutter/flutter/issues/55092) for Display P3 on iOS with plans to follow up with support for Android and then the framework itself.
+
+## Detailed design
+
+### JS Changes
+
+1. Parse the color() function in [normalizeColor](https://github.com/facebook/react-native/blob/63213712125795ac082597dad2716258b90cdcd5/packages/normalize-color/index.js#L235)
+
+```js
+const COLOR_SPACE = /display-p3|srgb/
+
+function getMatchers() {
+  if (cachedMatchers === undefined) {
+    cachedMatchers = {
+      color: new RegExp(
+        'color(' +
+          call(COLOR_SPACE, NUMBER, NUMBER, NUMBER) +
+          '|' +
+          callWithSlashSeparator(COLOR_SPACE, NUMBER, NUMBER, NUMBER, NUMBER) +
+          ')'
+      ),
+```
+
+2. Include the color space in the return value of [StyleSheet processColor](https://github.com/facebook/react-native/blob/63213712125795ac082597dad2716258b90cdcd5/packages/react-native/Libraries/StyleSheet/processColor.js) and [Animated processColor](https://github.com/facebook/react-native/blob/main/packages/react-native/Libraries/Animated/nodes/AnimatedColor.js)
+
+```js
+return { ["display-p3"]: true, r, g, b, a };
+```
+
+### iOS Changes
+
+1. Update [RCTConvert](https://github.com/facebook/react-native/blob/781b637db4268ad7f5f3910d99ebb5203467840b/packages/react-native/React/Base/RCTConvert.m#L881) to handle setting new color values for color space. If color space is not included preserve existing behavior.
```

Comment:
could be helpful to introduce a factory function:
```objc
enum ColorSpace: NSInteger {
  case sRGB = 0
  case displayP3 = 1
};

+ (UIColor *) createColorFrom:(CGFloat)red green:(CGFloat)green blue:(CGFloat)blue alpha:(CGFloat)alpha andColorSpace:(ColorSpace)colorSpace 
{
  if (colorSpace == displayP3) {
    return [UIColor colorWithDisplayP3Red:...];
  }
  return [UIColor red:...];
}
```
Can also be a UIColor category!

This should allow us to centralize where colors are created.
Of course, we need to find which pod is used by the three conversion files and include it from there.